oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/util/ElasticIndexUtils.java

```diff
@@ -31,6 +31,110 @@ public class ElasticIndexUtils {
 
     private static final Logger LOG = LoggerFactory.getLogger(ElasticIndexUtils.class);
 
+    /**
+     * Convert a JCR property name to a Elasticsearch field name.
+     * Notice that "|" is not allowed in JCR names.
+     *
+     * @param propertyName the property name
+     * @return the field name
+     */
+    public static String fieldName(String propertyName) {
+        if(propertyName.startsWith(":")) {
+            // there are some hardcoded field names
+            return propertyName;
+        }
+        String fieldName = propertyName;
+        boolean escape = false;
+        if (fieldName.isBlank()) {
+            // empty field name or field names that only consist of spaces
+            escape = true;
+        } else {
+            // 99.99% property names are OK,
+            // so we loop over the characters first
+            for (int i = 0; i < fieldName.length(); i++) {
+                switch (fieldName.charAt(i)) {
+                case '|':
+                case '.':
+                case '^':
+                case '_':
+                    escape = true;
+                }
+            }
+        }
+        if (escape) {
+            StringBuilder buff = new StringBuilder(fieldName.length());
+            for (int i = 0; i < fieldName.length(); i++) {
+                char c = fieldName.charAt(i);
+                switch (c) {
+                case '|':
+                    buff.append("||");
+                    break;
+                case '.':
+                case '^':
+                    buff.append('|').append(Integer.toHexString(c)).append('|');
```

Comment:
The characters here can only be `.` and `^`, so we could avoid the call to `toHexString(c)` and just append a constant string, like `|XXX|` and `|YYY|`. For instance:
```
                case '.':
                    buff.append("|XXX|"); break;
                case '^':
                     buff.append("|YYY|"); break;
```

Reply:
Yes, it's possible... but this is not on the critical path, and I wanted to allow extending the list if needed... it's really hard to get a good list... the `^` I found myself, and I first also had `#` and `-` and I'm not sure if those never need to be escaped...

Reply:
I'm wondering if this should not be considered as being on the critical path. For most indexes, this logic will never be executed, sure. But for the very few indexes that contain one or more properties with characters that need to be escaped, won't this logic be called very frequently when processing that specific index? In that case, there could be a noticeable impact in performance when reindexing or processing that index.